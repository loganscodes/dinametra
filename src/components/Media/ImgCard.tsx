import { Tooltip } from "react-tooltip"
import { ImgCardProps } from "../../interfaces/interfacesMedia"


const ImgCard = ({ notice, tooltipID, tooltipContent }: ImgCardProps) => {
    return (
        <>
            <div tabIndex={0} data-tooltip-id={tooltipID} data-tooltip-content={tooltipContent} className={`flex justify-center`} >
                {notice.links && notice.links.length > 0 ? (
                    <img src={notice.links[0].href} alt="" className='w-80 h-80 object-cover' loading='lazy' />
                ) : (
                    <img src='./NASA_logo.png' alt="" className='w-72 object-cover opacity-40' loading='lazy' />
                )}
            </div>
            <Tooltip id={tooltipID} />
        </>

    )
}

export default ImgCard