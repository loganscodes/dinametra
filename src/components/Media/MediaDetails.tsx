import { Item } from '../../interfaces/interfaces';
import UITitle from '../UI/UITitle';
import UIParagraph from '../UI/UIParagraph';
import SeeVideo from '../SeeVideo';
import UIButton from '../UI/UIButton';


interface Props {
    selectedNotice: Item;
    handleClose: () => void;
}

const MediaDetails = ({ selectedNotice, handleClose }: Props) => {

    return (
        <div className="fixed top-0 left-0 w-full h-full bg-white bg-opacity-75 flex justify-center items-center  overflow-y-auto">
            <div className="max-w-lg bg-white rounded-lg shadow-lg p-8">
                {selectedNotice.data.map((datum, index) => (
                    <div key={index}>

                        <UITitle tag='h2' className='text-2xl font-bold mb-4 text-center' tooltipID='title' title={datum.title} tooltipContent={datum.title} />
                        <UIParagraph tooltipID='description' tooltipContent='Description' className='mb-4 text-justify' content={datum.description} />
                        <SeeVideo selectedNotice={selectedNotice.href} />
                        <UIParagraph className='font-bold' content='Keywords' tooltipContent='Keywords' tooltipID='key' />
                        <UIParagraph className='text-blue-500' content={datum.keywords.join(', ')} useDangerousHTML={true} />

                    </div>
                ))}
                <UIButton onClick={handleClose} text='Close'/>
            </div>
        </div>
    );
};

export default MediaDetails;
