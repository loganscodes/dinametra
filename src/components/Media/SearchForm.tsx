import { Tooltip } from 'react-tooltip';

interface Props {
    handleSearch: (e: React.FormEvent<HTMLFormElement>) => void;
    inputValue: string;
    setInputValue: React.Dispatch<React.SetStateAction<string>>;
}

const SearchForm = ({ handleSearch, inputValue, setInputValue }: Props) => {
    return (
        <div className='my-10'>
            <form onSubmit={handleSearch} className='flex gap-5'>
                <input 
                    data-tooltip-id="search" data-tooltip-content='Can type Mars, Earth, Sun, etc...' data-tooltip-place="top"
                    type='text' 
                    value={inputValue} 
                    onChange={(e) => setInputValue(e.target.value)} 
                    placeholder='Search for Notices' 
                    className='p-2 border rounded'
                />
                <button type='submit' className='bg-blue-500 text-white p-2 px-5 rounded-lg'>Search</button>
                <Tooltip id="search" />
            </form>
            
        </div>
    );
};

export default SearchForm;
