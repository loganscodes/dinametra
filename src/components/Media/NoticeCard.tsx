import React from 'react';
import { Tooltip } from 'react-tooltip';
import UITitle from '../UI/UITitle';
import UIParagraph from '../UI/UIParagraph';
import ImgCard from './ImgCard';
import { NoticeCardsProps } from '../../interfaces/interfacesMedia';



const NoticeCard = ({ notice, handleNoticeClick }: NoticeCardsProps) => {

    const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
        if (event.key === 'Enter') {
            handleNoticeClick(notice);
        }
    };

    return (
        <div
            role="button"
            tabIndex={0}
            data-tooltip-id="card"
            data-tooltip-content='Click To See Info'
            data-tooltip-place="top"
            className={`border-4 border-gray-700 px-5 rounded-2xl bg-white  ${notice.links && notice.links.length > 0 ? 'cursor-pointer' : 'cursor-not-allowed'}`}
            onClick={notice.links && notice.links.length > 0 ? () => handleNoticeClick(notice) : undefined}
            onKeyDown={handleKeyDown}
        >
            <UITitle tag='h2' tooltipID='title-card' tooltipContent={notice.data[0].title} className='font-bold text-center my-5 h-12' title={notice.data[0].title} />

            <ImgCard notice={notice} tooltipID='img-card' tooltipContent='Image Card' />

            <UIParagraph tooltipID='date' tooltipContent={new Date(notice.data[0].date_created).toLocaleDateString()} className='text-center font-bold mt-5' content={new Date(notice.data[0].date_created).toLocaleDateString()} />
            <Tooltip id="card" />
        </div>
    );
};

export default React.memo(NoticeCard);
