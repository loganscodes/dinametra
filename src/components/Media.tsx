import React, { useEffect } from 'react'
import { useMedia } from '../hooks/useMedia'
import SpinerLoading from './SpinerLoading'
import SearchForm from './SearchForm'

const Media = () => {

    const { handleSearch, inputValue, setInputValue, selectedYear, handleYearChange, notices, loading, error, filterByYear, handleNoticeClick, selectedNotice, handleClose } = useMedia()

    useEffect(() => {
        const handleKeyDown = (event: KeyboardEvent) => {
            if (event.key === 'Escape') {
                handleClose();
            }
        };

        window.addEventListener('keydown', handleKeyDown);
        return () => {
            window.removeEventListener('keydown', handleKeyDown);
        };
    }, [handleClose]);

    return (
        <>

            <h1 className='text-center text-4xl font-bold uppercase mt-10'>Media</h1>

            <div className='flex flex-wrap items-center justify-around mb-10 lg:mb-10'>

                <SearchForm handleSearch={handleSearch} inputValue={inputValue} setInputValue={setInputValue}/>


                <div className='flex justify-center  items-center gap-5'>
                    <p className='font-semibold text-md'>Select Year to Filter</p>
                    <select value={selectedYear} onChange={handleYearChange} className='p-2 rounded-lg'>
                        <option value="">All years</option>
                        {Array.from(new Set(notices.map((notice) => new Date(notice.data[0].date_created).getFullYear()))).map((year) => (
                            <option key={year} value={year.toString()}>{year}</option>
                        ))}
                    </select>
                </div>

            </div>

            {
                loading &&
                <SpinerLoading />
            }

            {error && <div>Error...</div>}

            <div className='grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-3 mx-5'>
                {notices.filter(filterByYear).map((notice) => (
                    <div key={notice.data[0].nasa_id} className='border-2 border-red-500 px-5 rounded-2xl'>
                        <h2 className='font-bold text-center my-5 h-12'>{notice.data[0].title}</h2>
                        <div className='flex justify-center' onClick={() => handleNoticeClick(notice)}>
                            {notice.links && notice.links.length > 0 ? (
                                <img src={notice.links[0].href} alt="" className='cursor-pointer w-80 h-80 object-cover ' />
                            ) : (
                                <img src='./NASA_logo.png' alt="" className='cursor-pointer w-80 h-80 object-cover ' />
                            )}
                        </div>
                        <p className='text-center font-bold mt-5'>{new Date(notice.data[0].date_created).toLocaleDateString()}</p>
                    </div>
                ))}
            </div>


            {selectedNotice && (
                <div className="fixed top-0 left-0 w-full h-full bg-white bg-opacity-75 flex justify-center items-center">
                    <div className="max-w-lg bg-white rounded-lg shadow-lg p-8">
                        <h2 className="text-xl font-bold mb-4">{selectedNotice.data[0].title}</h2>
                        <p className="mb-4">{selectedNotice.data[0].description}</p>
                        <p className='font-bold'>Keywords</p>
                        <div className='flex'>
                            <p>{selectedNotice.data[0].keywords}</p>
                        </div>
                        <button onClick={handleClose} className="mt-4 bg-blue-500 text-white px-4 py-2 rounded-lg">Close</button>
                    </div>
                </div>
            )}




        </>
    )
}

export default Media