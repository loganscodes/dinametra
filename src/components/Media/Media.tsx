import React, { useEffect, Suspense } from 'react';
import { useMedia } from '../../hooks/useMedia';
import SearchForm from './SearchForm';
import SpinerLoading from '../UI/UISpinerLoading';
import UITitle from '../UI/UITitle';
import NoticeCard from './NoticeCard';
import UIErrorText from '../UI/UIErrorText';
import UISearchYear from '../UI/UISearchYear';
import { usePagination } from '../../hooks/usePagination';
import UIPagination from '../UI/UIPagination';

const MediaDetails = React.lazy(() => import('./MediaDetails'));

const Media = () => {
    
    const { handleSearch, inputValue, setInputValue, selectedYear, handleYearChange, notices, loading, error, filterByYear, handleNoticeClick, selectedNotice, handleClose } = useMedia();

    const { currentPage, totalPages, currentItems, handleNextPage, handlePreviousPage } = usePagination({notices, filterByYear})


    useEffect(() => {
        const handleKeyDown = (event: KeyboardEvent) => {
            if (event.key === 'Escape') {
                handleClose();
            }
        };

        window.addEventListener('keydown', handleKeyDown);
        return () => {
            window.removeEventListener('keydown', handleKeyDown);
        };
    }, [handleClose]);

    return (
        <>
            <UITitle tag='h1' className='text-2xl lg:text-4xl font-bold text-center my-10 text-white' tooltipID='title' title="Media from Nasa" tooltipContent='Media from Nasa' />

            <div className='flex flex-wrap items-center justify-around mb-10 lg:mb-10'>
                <SearchForm handleSearch={handleSearch} inputValue={inputValue} setInputValue={setInputValue} />
                <div className='flex justify-center items-center gap-5 '>
                    <UISearchYear selectedYear={selectedYear} onChange={handleYearChange} notices={notices} />
                </div>
            </div>

            {loading && <SpinerLoading />}
            {error && <UIErrorText />}

            <div className='grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-3 mx-5'>
                {currentItems.map((notice) => (
                    <NoticeCard key={notice.data[0].nasa_id} notice={notice} handleNoticeClick={handleNoticeClick} />
                ))}
            </div>

            <UIPagination onClickPrevPage={handlePreviousPage} currentPage={currentPage} totalPages={totalPages} onClickNextPage={handleNextPage} />

            {selectedNotice && (
                <Suspense fallback={<SpinerLoading />}>
                    <MediaDetails selectedNotice={selectedNotice} handleClose={handleClose} />
                </Suspense>
            )}
        </>
    );
};

export default Media;
