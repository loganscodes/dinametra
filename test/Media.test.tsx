import React from 'react';
import { render} from '@testing-library/react';
import Media from '../src/components/Media/Media';
import '@testing-library/jest-dom'

describe('Media component', () => {
    test('renders without errors', () => {
        const { container } = render(<Media />);
        expect(container).toMatchSnapshot();
    });
});