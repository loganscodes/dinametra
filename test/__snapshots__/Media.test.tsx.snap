// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media component renders without errors 1`] = `
<div>
  <h1
    aria-label="Media from Nasa"
    class="text-2xl lg:text-4xl font-bold text-center my-10 text-white"
    data-tooltip-content="Media from Nasa"
    data-tooltip-id="title"
    data-tooltip-place="top"
    tabindex="0"
  >
    Media from Nasa
  </h1>
  <div
    class="flex flex-wrap items-center justify-around mb-10 lg:mb-10"
  >
    <div
      class="my-10"
    >
      <form
        class="flex gap-5"
      >
        <input
          aria-label="moon"
          class="p-2 border-4 rounded-xl border-gray-700"
          data-tooltip-content="Can type Mars, Earth, Sun, etc..."
          data-tooltip-id="search"
          data-tooltip-place="top"
          placeholder="Search media"
          type="text"
          value="moon"
        />
        <button
          aria-label="Search"
          class=" bg-gray-700 text-white px-4 py-2 rounded-lg"
          data-tooltip-content="Search Button"
          data-tooltip-id="search-button"
          data-tooltip-place="top"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="flex justify-center items-center gap-5 "
    >
      <p
        class="font-semibold text-md text-white"
      >
        Select Year to Filter
      </p>
      <select
        class="p-2 rounded-lg border-4 border-gray-700"
        data-tooltip-content="Search by year"
        data-tooltip-id="searchyear"
        data-tooltip-place="top"
      >
        <option
          value=""
        >
          All years
        </option>
      </select>
    </div>
  </div>
  <div
    aria-label="Network Down, please retry later"
    class=" flex flex-col items-center justify-center h-96"
  >
    <p
      class="text-white text-center text-4xl font-bold"
    >
      Network Down, please retry later
    </p>
  </div>
  <div
    class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-3 mx-5"
  />
  <div
    class="flex justify-center my-5"
  >
    <button
      class="px-4 py-2 bg-gray-700 text-white rounded-l-lg"
      disabled=""
    >
      Previous
    </button>
    <span
      class="px-4 py-2 bg-gray-300 text-gray-700"
    >
      Page 
      1
       of 
      0
    </span>
    <button
      class="px-4 py-2 bg-gray-700 text-white rounded-r-lg"
    >
      Next
    </button>
  </div>
</div>
`;
